File: src/library/scala/runtime/LambdaDeserialize.java
package scala.runtime;


import java.lang.invoke.*;
import java.util.Arrays;
import java.util.HashMap;

public final class LambdaDeserialize {
    private LambdaDeserialize() {
    }

    public static CallSite bootstrap(MethodHandles.Lookup lookup, String invokedName,
                                     MethodType invokedType) throws Throwable {
        HashMap<String, MethodHandle> cache = new HashMap<>();
        LambdaDeserializer$ l = LambdaDeserializer$.MODULE$;
        MethodType type = MethodType.fromMethodDescriptorString("(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/util/Map;Ljava/lang/invoke/SerializedLambda;)Ljava/lang/Object;", lookup.getClass().getClassLoader());
        MethodHandle deserializeLambda = lookup.findVirtual(LambdaDeserializer$.class, "deserializeLambda", type);
        MethodHandle bound = deserializeLambda.bindTo(l).bindTo(lookup).bindTo(cache);
        MethodHandle exact = bound.asType(invokedType);
        return new ConstantCallSite(exact);
    }
}

File: src/library/scala/runtime/LambdaDeserialize.scala
package scala.runtime

import java.lang.invoke._
import java.util

import scala.annotation.varargs
import scala.collection.mutable
object LambdaDeserialize {
  @throws[Throwable]
  @varargs
  def bootstrap(lookup: MethodHandles.Lookup, invokedName: String, invokedType: MethodType, targetMethods: MethodHandle*): CallSite = {
    val exact = MethodHandleConstants.LAMBDA_DESERIALIZE_DESERIALIZE_LAMBDA.bindTo(new LambdaDeserialize(lookup, targetMethods.asInstanceOf[mutable.WrappedArray[MethodHandle]].array)).asType(invokedType)
    new ConstantCallSite(exact)
  }
  def nameAndDescriptorKey(name: String, descriptor: String): String = name + descriptor
}

final class LambdaDeserialize private(var lookup: MethodHandles.Lookup, val targetMethods: Array[MethodHandle]) {
  private var targetMethodMap: util.HashMap[String, MethodHandle] = new util.HashMap[String, MethodHandle](targetMethods.length)

  for (targetMethod <- targetMethods) {
    val info = lookup.revealDirect(targetMethod)
    val key = LambdaDeserialize.nameAndDescriptorKey(info.getName, info.getMethodType.toMethodDescriptorString)
    targetMethodMap.put(key, targetMethod)
  }

  final private val cache = new util.HashMap[String, MethodHandle]

  def deserializeLambda(serialized: SerializedLambda): Any = LambdaDeserializer.deserializeLambda(lookup, cache, targetMethodMap, serialized)
}
